CIP-0XXX/README.md

```diff
@@ -0,0 +1,561 @@
+| StIdle | MsgRequestMessageIdsNonBlocking | ack,req | StMessageIdsNonBlocking |
+| StMessageIdsNonBlocking | MsgReplyMessageIds | [(id,size)] | StIdle |
+| StIdle | MsgRequestMessages | [ids] | StMessages |
+| StMessages | MsgReplyMessages | [messages] | StIdle |
+| StMessageIdsBlocking | MsgDone |  | StDone |
+
+#### CDDL encoding specification
+
+
+```cddl
+ 1
+ 2  messageSubmissionMessage
+ 3    = msgInit
+ 4    / msgRequestMessageIds
+ 5    / msgReplyMessageIds
+ 6    / msgRequestMessages
+ 7    / msgReplyMessages
+ 8    / msgDone
+ 9 
+10  msgInit = [0]
+11  msgRequestMessageIds = [1, isBlocking, messageCount, messageCount]
+12  msgReplyMessageIds = [2, [ *messageIdAndSize ] ]
+13  msgRequestMessages = [3, messageIdList ]
+14  msgReplyMessages = [4, ]
+15  msgDone = [5, ]
+16
+17  isBlocking = false / true
+18  messageCount = word16
+19  messageId = bstr
+20  messageBody = bstr
+21  messageIdAndSize = [ messageId, messageSizeInBytes ]
+22  messageIdList = [ * messageId ]
+23  messageList = [ * message ]
+24  messageSizeInBytes = word32
+25  kesSignature = bstr
+26  operationalCertificate = bstr
+27  blockNumber = word32
+28  
+29  message = [
+30    messageId,
+31    messageBody,
+32    blockNumber,
+33    kesSignature,
+34    operationalCertificate
+35  ]
+30
+```
+
+### Client and server implementation
+
+This mini-protocol is designed with two goals in mind:
+* diffuse messages with high efficiency
+* protect from asymmetric resource atatckes from the message consumer against the message provider
+
+The mini-protocol is based on two pull-based operations:
+- the message consumer asks for message ids,
+- and uses these ids to request a batch of messages (which it has not received yet)
+
+The client is responsible for initiating the mini-protocol with a peer node, but the server (i.e. the other node) is the one who asks for information.
+
+The client maintains a limited size FIFO queue of outstanding messages for each of the servers it is connected to, so does the server with a mirror FIFO queue of message ids:
+- the server asks for the next message ids and acknowledges for the previous message ids received (and removed from its queue).
+- the client removes the acknowledged ids from the FIFO queue it maintains for the server.
+- the server can download the content of the messages by giving an unordered list of ids to the client.
+
+The protocol supports blocking and non-blocking requests:
+- the client must reply immediately to a non-blocking request.
+- the server must wait until the client has at least one message available.
+- if the current queue of the server is empty, it must use a blocking request and a non-blocking request otherwise.
+
+### Network load
+
+#### Mithril extra network usage
+
+> [!NOTE]
+> The computation must apply a multiplicative factor based on the number of peers connected and/or the overall redundancy of messages transmitted with the diffusion mechanism. The current value is **2**.
```

Comment:
Which computation refers this to?

Reply:
I explained to which computations the factor is applied.